### PR TITLE
[fix](multicast) fix DCHECK failure of block mem reuse for multicast

### PR DESCRIPTION
**Pull request by jacktengg**
## Proposed changes

Issue Number: close #xxx

failure message:
```
F20231025 16:39:08.590188 3535510 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1)  
*** Check failure stack trace: *** 
F20231025 16:39:08.591538 3535515 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1)                                                                                                                  
*** Check failure stack trace: *** 
F20231025 16:39:08.591538 3535515 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1) F20231025 16:39:08.592170 3535519 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1)  
*** Check failure stack trace: *** 
F20231025 16:39:08.591538 3535515 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1) F20231025 16:39:08.592170 3535519 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1) F20231025 16:39:08.592965 3535523 block.cpp:694] Check failed: d.column->use_count() == 1 (2 vs. 1)  
*** Check failure stack trace: *** 
    @     0x55770c4767e6  google::LogMessage::SendToLog()
    @     0x55770c4767e6  google::LogMessage::SendToLog()
    @     0x55770c4767e6  google::LogMessage::SendToLog()
    @     0x55770c4767e6  google::LogMessage::SendToLog()
    @     0x55770c472db0  google::LogMessage::Flush()
    @     0x55770c472db0  google::LogMessage::Flush()
    @     0x55770c472db0  google::LogMessage::Flush()
    @     0x55770c472db0  google::LogMessage::Flush()
    @     0x55770c477029  google::LogMessageFatal::~LogMessageFatal()
    @     0x55770c477029  google::LogMessageFatal::~LogMessageFatal()
    @     0x55770c477029  google::LogMessageFatal::~LogMessageFatal()
    @     0x55770c477029  google::LogMessageFatal::~LogMessageFatal()
    @     0x5576f438a4ad  doris::vectorized::Block::clear_column_data()
    @     0x5576f438a4ad  doris::vectorized::Block::clear_column_data()
    @     0x5576f438a4ad  doris::vectorized::Block::clear_column_data()
    @     0x5576f438a4ad  doris::vectorized::Block::clear_column_data()
    @     0x55770c2b4412  doris::pipeline::PipelineTask::execute()
    @     0x55770c2b4412  doris::pipeline::PipelineTask::execute()
    @     0x55770c2b4412  doris::pipeline::PipelineTask::execute()
    @     0x55770c2b4412  doris::pipeline::PipelineTask::execute()
    @     0x55770c442b8f  doris::pipeline::TaskScheduler::_do_work()
    @     0x55770c442b8f  doris::pipeline::TaskScheduler::_do_work()


0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /mnt/disk2/tengjianping/doris-1.1/be/src/common/signal_handler.h:417
 1# 0x00007F4120917400 in /lib64/libc.so.6
 2# __GI_raise in /lib64/libc.so.6
 3# __GI_abort in /lib64/libc.so.6
 4# 0x000055770C480FFD in /mnt/disk2/tengjianping/doris-1.1/output/be2/lib/doris_be
 5# 0x000055770C47330A in /mnt/disk2/tengjianping/doris-1.1/output/be2/lib/doris_be
 6# google::LogMessage::SendToLog() in /mnt/disk2/tengjianping/doris-1.1/output/be2/lib/doris_be
 7# google::LogMessage::Flush() in /mnt/disk2/tengjianping/doris-1.1/output/be2/lib/doris_be
 8# google::LogMessageFatal::~LogMessageFatal() in /mnt/disk2/tengjianping/doris-1.1/output/be2/lib/doris_be
 9# doris::vectorized::Block::clear_column_data(int) in /mnt/disk2/tengjianping/doris-1.1/output/be2/lib/doris_be
10# doris::pipeline::PipelineTask::execute(bool*) at /mnt/disk2/tengjianping/doris-1.1/be/src/pipeline/pipeline_task.cpp:285
11# doris::pipeline::TaskScheduler::_do_work(unsigned long) at /mnt/disk2/tengjianping/doris-1.1/be/src/pipeline/task_scheduler.cpp:264
12# void std::__invoke_impl<void, void (doris::pipeline::TaskScheduler::*&)(unsigned long), doris::pipeline::TaskScheduler*&, unsigned long&>(std::__invoke_memfun_deref, void (doris::pipeline::TaskScheduler::*&)(unsigned long), doris::pipeline::TaskScheduler*&, unsigned long&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:74
13# std::__invoke_result<void (doris::pipeline::TaskScheduler::*&)(unsigned long), doris::pipeline::TaskScheduler*&, unsigned long&>::type std::__invoke<void (doris::pipeline::TaskScheduler::*&)(unsigned long), doris::pipeline::TaskScheduler*&, unsigned long&>(void (doris::pipeline::TaskScheduler::*&)(unsigned long), doris::pipeline::TaskScheduler*&, unsigned long&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
14# void std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:420
15# void std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>::operator()<, void>() at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:503
16# void std::__invoke_impl<void, std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>&>(std::__invoke_other, std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
17# std::enable_if<is_invocable_r_v<void, std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>&>, void>::type std::__invoke_r<void, std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>&>(std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)>&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:117
18# std::_Function_handler<void (), std::_Bind<void (doris::pipeline::TaskScheduler::*(doris::pipeline::TaskScheduler*, unsigned long))(unsigned long)> >::_M_invoke(std::_Any_data const&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
19# std::function<void ()>::operator()() const at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
20# doris::FunctionRunnable::run() at /mnt/disk2/tengjianping/doris-1.1/be/src/util/threadpool.cpp:48
21# doris::ThreadPool::dispatch_thread() at /mnt/disk2/tengjianping/doris-1.1/be/src/util/threadpool.cpp:543
22# void std::__invoke_impl<void, void (doris::ThreadPool::*&)(), doris::ThreadPool*&>(std::__invoke_memfun_deref, void (doris::ThreadPool::*&)(), doris::ThreadPool*&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:74
23# std::__invoke_result<void (doris::ThreadPool::*&)(), doris::ThreadPool*&>::type std::__invoke<void (doris::ThreadPool::*&)(), doris::ThreadPool*&>(void (doris::ThreadPool::*&)(), doris::ThreadPool*&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
24# void std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:420
25# void std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>::operator()<, void>() at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:503
26# void std::__invoke_impl<void, std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&>(std::__invoke_other, std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
27# std::enable_if<is_invocable_r_v<void, std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&>, void>::type std::__invoke_r<void, std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&>(std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:117
28# std::_Function_handler<void (), std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()> >::_M_invoke(std::_Any_data const&) at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
29# std::function<void ()>::operator()() const at /mnt/disk2/tengjianping/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
30# doris::Thread::supervise_thread(void*) at /mnt/disk2/tengjianping/doris-1.1/be/src/util/thread.cpp:494
31# start_thread in /lib64/libpthread.so.0
32# __GI___clone in /lib64/libc.so.6
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run p0

File: be/src/vec/exprs/vexpr_context.cpp [hunk 1]
```diff
@@ -288,14 +288,21 @@ Status VExprContext::execute_conjuncts_and_filter_block(const VExprContextSPtrs&
...
+    auto rows = input_block.rows();
     vectorized::Block tmp_block(input_block.get_columns_with_type_and_name());
     vectorized::ColumnsWithTypeAndName result_columns;
     for (auto& vexpr_ctx : output_vexpr_ctxs) {
         int result_column_id = -1;
         RETURN_IF_ERROR(vexpr_ctx->execute(&tmp_block, &result_column_id));
         DCHECK(result_column_id != -1);
-        result_columns.emplace_back(tmp_block.get_by_position(result_column_id));
+        const auto& col = tmp_block.get_by_position(result_column_id);
+        if (do_projection) {
```
Comment (yiguolei) [inline be/src/vec/exprs/vexpr_context.cpp:301]: may have performance issues

File: be/src/vec/exprs/vexpr_context.cpp [hunk 2]
```diff
@@ -288,14 +288,21 @@ Status VExprContext::execute_conjuncts_and_filter_block(const VExprContextSPtrs&
 }
 
 Status VExprContext::get_output_block_after_execute_exprs(
-        const VExprContextSPtrs& output_vexpr_ctxs, const Block& input_block, Block* output_block) {
+        const VExprContextSPtrs& output_vexpr_ctxs, const Block& input_block, Block* output_block,
```
Comment (yiguolei) [inline be/src/vec/exprs/vexpr_context.cpp:291]: add more comment to explain why we need this parameter

**Comment by jacktengg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by mrhhsg — APPROVED** [review]
LGTM

**Pull request merged**